### PR TITLE
Entropy initial temperature lookup (WIP)

### PR DESCRIPTION
**Pull request by lhy11009**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

In this PR, I create a new model in the initial composition to convert the initial temperature to the initial entropy by looking up a table. In the table, the pressure and the temperature need to be assigned as X and Y axis, respectively, and the entropy needs to be looked up.
After this PR, one could provide an initial temperature field but use the entropy method for computation.
This is currently working and near completion.
I want to ask for some help and review from @jdannberg, @RanpengLi,@bobmyhill before wrapping up (listed below).
FYI @mibillen

### What I have done:

- A new "entropy table lookup" model in the initial composition

- A new test entropy_initial_lookup

This test takes a constant temperature field (left subfigure) and converts it to an entropy field decreasing with depth (right subfigure).
combine_images (3)

### What I need help with before finishing:

- Discuss the implementation during user meetings.

- A consistent table that converts (pressure, entropy) -> temperature.

I am playing with a pyrolite composition. The model I generated from Perple_X is (temperature, pressure)-> entropy, so I need this one additional table to check the consistency of this approach (that I could get the same temperature from entropy).

- NaN values in the table

The values I have from Perple_X have NaN values. Currently, I assign them to a big value (10000.0), I'd like to know whether this is a good option.

### Other things I want to complete

- Add another test that uses the WorldBuilder as the initial temperature model and generates an initial slab (same as used for the phase transition cookbook).

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

**Comment by gassmoeller** [comment]
Hi @lhy11009, the idea sounds interesting and we can discuss details during a user meeting. However, my initial question would be: Why do you need this plugin in addition to the ascii_data plugin? Or phrased differently: Could the ascii_data plugin not do already what you want to do?

**Comment by lhy11009** [comment]
@gassmoeller That's indeed an important question to think through. It would be very helpful to clarify what we want to achieve here.
If I understand it right, the ascii_data plugin will specify a composition field. In this case of application, we could use it for an entropy field. That's indeed very helpful. Moreover, I think what you mean is to have an initial entropy field (e.g. a slab) and use the ascii_data plugin, is that right? 
If that what's the question contains, the answer is both yes and no. It could technically cover what I want to model but it's not going to achieve the same level of neatness this PR seems to provide.
The idea of this PR and the new plugin is more about changing the workflow. The new workflow will be using the WorldBuilder for the initial condition and have the (T, P) -> entropy table convert it into an initial entropy field. 
In other words, we are trying to wrap up the entropy method as a black box for the user. Provided two (or two sets of) lookup tables, (T,P) -> S and (P, ) -> T, are given, the user would proceed the same way as if the entropy method is not used.
Please tell me what you think.

**Comment by lhy11009** [comment]
`set Initial entropy from temperature field = True`
This is a good idea, I could easily make the change.

> we should require input in SI units 
For the current version of the PR, I copied a bit from the plugin in the entropy_adiabat benchmark and used their convention. So I could make the change after a discussion with @jdannberg and @RanpengLi 

 
> I already find it unintuitive to have Entropy model as a Material model rather than as a different model class.
I agree, if this could be somehow managed, it will make the parameters I need in the PR more intuitive (don't need to be in the composition section)

I'll try to participate in the next user meeting (08/16) and discuss.

**Comment by gassmoeller** [comment]
@lhy11009 I will leave a more detailed comment later, just wanted to let you know I cannot make the user meeting tomorrow. We could either meet separately, or during the next user meeting, whichever you prefer.

File: doc/modules/changes/20230811_haoyuan [hunk 1]
```diff
@@ -0,0 +1,7 @@
+ New: There is now an entropy table lookup model
+ in the initial composition model.
+ This model takes an initial temperature field and
+ converts it to an initial entropy field by looking
+ up in a table.
```
Comment (gassmoeller) [inline doc/modules/changes/20230811_haoyuan:5]:
```suggestion
 converts it to an initial entropy field using
 a lookup table.
```

File: source/initial_composition/entropy_table_lookup.cc [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/initial_composition/entropy_table_lookup.h>
+#include <aspect/initial_temperature/interface.h>
+#include <aspect/adiabatic_conditions/interface.h>
+#include <aspect/material_model/interface.h>
+#include <aspect/melt.h>
```
Comment (gassmoeller) [inline source/initial_composition/entropy_table_lookup.cc:26]: I dont think you need the last two includes.

File: source/initial_composition/entropy_table_lookup.cc [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        {
+          double temperature = initial_temperature_manager->initial_temperature(position);
+          double pressure = this->get_adiabatic_conditions().pressure(position);
+
+          // Convert pressure from Pa to bar, bar is used in the table.
+          Point<2> temperature_pressure(temperature, pressure / 1.e5);
+
+          double entropy = material_lookup->get_data(temperature_pressure, 0);
+          // double entropy = 2 * temperature;
+          return entropy;
```
Comment (gassmoeller) [inline source/initial_composition/entropy_table_lookup.cc:80]:
```suggestion
          const double entropy = material_lookup->get_data(temperature_pressure, 0);
          return entropy;
```

File: tests/entropy_initial_lookup.prm [hunk 4]
```diff
@@ -0,0 +1,145 @@
+#
+
+set Additional shared libraries            = $ASPECT_SOURCE_DIR/benchmarks/entropy_adiabat/plugins/libentropy_plugins.so
```
Comment (gassmoeller) [inline tests/entropy_initial_lookup.prm:3]: This line will not work in the test. If you indeed need those plugins for this test, you need to instead add a .cc file with the same name as the test, and include all .cc files that you need inside that file. (see entropy_adiabat.cc in the same folder).

**Review by gassmoeller — COMMENTED** [review]
Some comments on the implementation, generally this looks good. Forget my comment about using the `EntropyReader` class from #5341, you need the inverse operation here, and your implementation is correct for that.

As you mentioned and as we discussed today the main challenge for you right now is that to test your implementation (and to run practical models) you need tables that allow you to go from temperature to entropy and back to temperature for the same composition. Let me know when you have figured out a good testcase then we can proceed with this PR.

**Review by lhy11009 — COMMENTED** [review]
@gassmoeller. I have run into an issue in testing. I have put the plugins I need into a separate .cc file that carries the name of the test, but the run fails because of it. It seems that the compilation of that .cc file is wrong without locating the right mpi.h file. I took a step back and ran the entropy_adiabat test and it produced the same error. Here is the error output when running the entropy_adiabat test (aspect/build_entropy_initial_temperature/Testing/Temporary/LastTest.log). 
[LastTest.log](https://github.com/geodynamics/aspect/files/12424794/LastTest.log)

Can you help me figure out what the issue is?

**Comment by lhy11009** [comment]
@gassmoeller , I think this PR is already in good shape.  We could proceed and start the reviewing. There is one remaining issue I need help:

- run the test with the .cc file with the same name (see above).

Maybe we can meet in the next user meeting.

**Comment by gassmoeller** [comment]
> Can you help me figure out what the issue is?

I am not sure what the error on your system is, but the one on the reference tester looks different (see https://github.com/geodynamics/aspect/actions/runs/6065701057/job/16455501549?pr=5360#step:6:607). I think you have a conflict with some changes that happened in #5365 and one of the files you include in your test is no longer a separate plugin but instead moved into main aspect. Can you rebase your PR, remove the include line that causes the error, and try again?

**Comment by lhy11009** [comment]
@gassmoeller You are right, there is some different between mine and the tester machine. I rebased my branch and it doesn't resolve this issue. The running of the entropy_adiabat test still crushes with the same error message. I have to switch to using the docker image for generating test results. The running of the entropy_adiabat succeeds within the docker image, demostrating the issue is indeed the difference from the tester machine.

**Comment by gassmoeller** [comment]
> Is the file in the docker somehow linked to the file in my local directory. In this case, if I adjust the contents in my test prm file, is the change updated on the docker side as well? Or should I finalize my test prm file before running the docker.

Files outside and inside the container are linked if you followed the instructions (i.e. if you use `docker run [...] -v [...]:/opt/aspect` then `/opt/aspect` inside the container and your aspect directory outside of the container are the same directory).

>    If I run the "ASPECT_GENERATE_REFERENCE_OUTPUT=1 ctest -j 4" command to generate the files in the docker, would that update my local test files?

Yes.


Based on the files you sent I cannot see why the test on your system would fail, but would work inside the container. One thing you can try is deleting the tests/ folder in the build directory, then rerun `make setup_tests` and rerun the test. Maybe it helps. If not we can meet sometime next week (best Wednesday or Thursday) and you can show me how you run the test.

**Comment by lhy11009** [comment]
@gassmoeller I indented the files and fix the test outputs.

File: source/initial_composition/entropy_table_lookup.cc [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+// explicit instantiations
+namespace aspect
+{
+  namespace InitialComposition
+  {
+    ASPECT_REGISTER_INITIAL_COMPOSITION_MODEL(EntropyTableLookUp,
+                                              "entropy table lookup",
+                                              "A class that implements initial conditions for the entropy field "
+                                              "by converting the initial temperature file through a look up table"
```
Comment (gassmoeller) [inline source/initial_composition/entropy_table_lookup.cc:130]:
```suggestion
                                              "by converting the initial temperature field through a look up table"
```

File: tests/entropy_initial_lookup.prm [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    set Material file name               = material_table_entropy_pressure.txt
+    set Lateral viscosity file name      = constant_lateral_vis_prefactor.txt
+    set Thermal conductivity formulation = constant
+    set Thermal conductivity             = 0.1
+  end
+
+  subsection Depth dependent viscosity
+    set Data directory                   = $ASPECT_SOURCE_DIR/benchmarks/entropy_adiabat/data/
+    set Data file name                   = constant_viscosity.txt
+  end
```
Comment (gassmoeller) [inline tests/entropy_initial_lookup.prm:130]: I think this part is not used, correct?

File: tests/entropy_initial_lookup_wb.wb [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+          "coupling depth": 50000.0,
+          "taper distance": 100000.0,
+          "min distance slab top": -100000.0,
+          "max distance slab top": 150000.0
+        }
+      ]
+    }
+  ],
+  "potential mantle temperature": 1573.0
+}
```
Comment (gassmoeller) [inline tests/entropy_initial_lookup_wb.wb:185]: missing newline at end of file.

**Review by gassmoeller — COMMENTED** [review]
Apologies for the long delay to review :-(. Very nice functionality, almost ready to merge. Please run make indent one more time and address the small comments I left.

File: tests/entropy_initial_lookup.prm (hunk 4 above)
Comment (lhy11009) [reply tests/entropy_initial_lookup.prm]: Fixed

File: tests/entropy_initial_lookup_wb.wb (hunk 7 above)
Comment (lhy11009) [reply tests/entropy_initial_lookup_wb.wb:185]: fixed

File: tests/entropy_initial_lookup.prm (hunk 6 above)
Comment (lhy11009) [reply tests/entropy_initial_lookup.prm:130]: Yes, this part is not needed. I have removed them from the two tests and remade the outputs.

**Comment by lhy11009** [comment]
@gassmoeller, I have addressed the comments and made indent.

**Review by gassmoeller — APPROVED** [review]
